#13 feat(LocalAuth) make LocalAuth Sign-up and Sign-in feature

uridonget opened the pull request:
# feat(LocalAuth) make LocalAuth Sign-up and Sign-in feature  \#13
## Overview
로컬 로그인 기능을 구현하였습니다.

<!---- Resolves: #(Isuue Number) -->
## PR Type

1. LocalAuth 테이블 추가
2. LocalAuth 토큰 발급 함수 추가 (get_local_auth_token)
3. LocalAuth ID 중복 확인 함수 추가 (check_local_auth_id)
4. LocalAuth 회원가입 함수 추가(local_auth_sign_up)

아직 LocalAuth 회원가입 과정에서 이메일로 확인하는 API 및 로직은 추가하지 않았습니다.
해당 기능은 2FA를 구현할 때 같이 구현하도록 하겠습니다.

- [ ] feat
- [ ] fix

## PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.

uridonget committed:
#13 feat(database): LocalAuth 테이블 추가

account 앱의 models.py에 LocalAuth 모델 추가.
admin 페이지에서 DB를 관리할 수 있도록 admin.py 파일에 LocalAuth 클래스를 추가.

uridonget committed:
#13 feat(LocalAuth): LocalAuth 토큰 발급 함수 추가.

1. LocalAuth 인증 후 토큰 발급 함수 추가
    - 위치 : backend/account/views.py
    - Request의 바디에서 id와 password를 추출
    - LocalAuth 테이블에서 입력받은 id와 password를 동시에 가진 유저를 탐색
    - 일치하는 객체가 있는 경우 해당 유저를 통해 JWT 생성
    - 생성된 JWT를 클라이언트의 쿠키에 저장 후 기본 페이지(http://localhost)로 리디렉션
2. LocalAuth get_local_auth_token 함수를 실행하는 엔드포인트 추가
    - 위치 : backend/account/urls.py

uridonget committed:
#13 feat(LocalAuth): check_local_auth_id 함수 추가

LocalAuth 회원가입 과정에서 필요한 id 중복확인(Figma:LoginPage-회원가입 참조) 함수 추가

1. LocalAuth 테이블에서 특정 ID의 중복여부를 확인하는 함수 추가
    - 위치 : backend/account/views.py
    - URL의 쿼리 스트링에서 local_id 추출
    - LocalAuth 테이블에서 입력받은 local_id와 중복되는 localID를 가진 유저를 탐색
    - 중복 여부를 JSON 형태로 리턴
2. LocalAuth check_local_auth_id 함수를 실행하는 엔드포인트 추가
    - 위치 : backend/account/urls.py

uridonget committed:
#13 feat(LocalAuth): local_auth_sign_up 함수 추가

LocalAuth 회원가입 함수 추가

1. 새로운 LocalAuth 객체를 생성하고 그와 연결된 User 객체를 생성하는 함수 추가
    - 위치 : backend/account/views.py
    - Request의 바디에서 local_id, local_password, user_email을 추출
    - LocalAuth 테이블에서 입력받은 local_id와 중복되는 localID를 가진 객체를 탐색
    	- 아이디가 중복된 객체가 존재한다면 에러메시지와 상태코드를 JSON 형태로 리턴
	- 아이디가 중복된 객체가 존재하지 않는다면 LocalAuth, User 객체를 생성 후 성공메시지와 상태코드를 JSON 형태로 리턴

2. LocalAuth local_auth_sign_up 함수를 실행하는 엔드포인트 추가
    - 위치 : backend/account/urls.py

3. get_local_auth_token 함수의 리턴값 수정
    - 위치 : backend/account/views.py
    - 기존에 HttpResponseRedirect 를 사용하여 기본페이지로 이동하던 것을 성공메시지를 리턴하는 JSON 형태로 변경
    - 프론트에서 처리하는 것이 구조적으로 더 좋을 것이라 판단하였음.

4. check_local_auth_id 함수의 리턴값 수정
    - 위치 : backend/account/views.py
    - 중복된 아이디가 존재하는 경우의 리턴값을 200에서 409(Conflict)로 수정하였음.

uridonget committed:
#13 fix(Black): Black 포맷에 맞도록 수정

기존의 코드에 Black formatter를 실행하였습니다.

allorak333 commented on backend/account/views.py:
```diff
@@ -170,3 +170,144 @@ def get_or_create_user_oauth(intra_id, user_email, user_image_path):
+        random_nickname = generate_random_nickname()
+        user = User.objects.create(
+            email=user_email,
+            imagePath="/static/image/default.jpeg",
```
oauth로 가입된 유저의 경우 image 경로가 42 서버의 cdn 서버로 설정되지만 localauth의 경우 imagePath에 서버 주소까지 포함되어있지 않아 실제 사용 시 혼란이 발생 할 것 같습니다.
환경변수를 이용하여 경로 앞에 서버 주소도 포함하는 로직이 필요해보입니다.

uridonget replied:
동의합니다 수정하겠습니다

allorak333 commented on backend/account/models.py:
```diff
@@ -45,3 +45,23 @@ class OAuth(models.Model):
+    user = models.OneToOneField(
+        User, on_delete=models.CASCADE, related_name="local_auth"
+    )
+    localId = models.CharField(max_length=30, unique=True, db_index=True)
```
index 설정은 검색 속도를 향상시키지만 상황에 따라 반대로 성능을 저하시키는 단점도 존재합니다.
따라서 localId 속성에 primay_key 속성 말고 db_index로 설정해놓은 이유가 필요합니다.

uridonget replied:
곰곰히 생각해보니 pk로 두지 않을 이유가 없는 것 같습니다! localId 뿐 아니라 OAuth 의 intraId 또한 pk로 두도록 수정하겠습니다.

uridonget committed:
#13 fix(LocalAuth): PR 커맨드 피드백에 맞추어 수정

1. LocalAuth 클래스 수정
    - 위치 : backend/account/models.py
    - localId를 Primary key로 설정하였습니다.

2. OAuth 클래스 수정
    - 위치 : backend/account/models.py
    - intraId를 Primary key로 설정하였습니다.

3. LocalAuth 회원가입 시 유저의 기본 이미지 url 수정
    - 위치 : backend/account/views.py
    - 기본 이미지를 /static/image/default.jpeg에서 http://localhost/static/image/default.jpeg로 수정

4. LocalAuth 비밀번호 관련하여 보안 향상
    - 위치 : backend/account/views.py
    - 비밀번호를 저장할 때 PBKDF2 알고리즘을 사용해서 해싱해서 저장.

allorak333 reviewed (approved):
LGTM!
